### ISSUE TITLE
cuda::NeuralNetwork is a stale fork of ai::NeuralNetwork and carries three bugs that have since been fixed

### ISSUE THREAD

**Issue by xoloki** [xoloki/jlib#138]
`jlib/cuda/neural.hh` is a 399-line copy of `jlib/ai/neural.hh` with the
multiplies routed through cuBLAS. It has not tracked the original, and four
branches of fixes have gone into the original without reaching it.

## What it is missing

**Fan-out initialisation (#131).**

```cpp
T hbound = pow(m_nhidden[0], -0.5);   // fan-in is m_ninput
T obound = pow(m_noutput,    -0.5);   // fan-in is m_nhidden.back()
```

Measured on the original at 4.6 points of MNIST accuracy at two hidden layers.

**Backpropagation through already-updated weights (#130).**

```cpp
gemm(..., m_who);              // m_who updated here
math::matrix<T> deep = m_who;  // and read here
```

Measured on the original as the difference between chance and 70% at four
hidden layers, and 88% to 96% at three.

**No mean gradient.** The step is `m_lrate * (...)` undivided by the batch, so
the effective learning rate scales with batch size -- a batch of 64 takes a 64x
step. The original divides, which is a no-op at a batch of one and correct
above it.

It also predates the selectable activation, so it is sigmoid-only with the
derivative written out by hand as `out ^ (1 - out)`.

## Why this is filed rather than fixed

It is gated on `--with-cuda` and does not build on the machine the fixes were
made on, so porting them would be three changes to numerical code that cannot
be compiled, let alone run or compared against a dataset. Three guesses in a
row is worse than an accurate issue.

## The better ending

Work is under way on an abstract backend in `jlib/ai`: an opaque tensor plus
the operations a training step needs, with a CPU implementation as the default
and `jlib/metal` supplying another. If a cuBLAS backend is written to the same
interface, this file becomes deletable rather than repairable, and the
divergence cannot happen again because there would be one copy of the training
step.

That is the argument for the backend design generally, and this file is the
evidence for it: forking the network for a second device produced a version
that silently missed every subsequent fix, and nothing noticed because it does
not build in CI.
